mmdata.c

```diff
@@ -1101,6 +1101,39 @@ int nmbrEq(nmbrString *s,nmbrString *t)
 }
 
 
+/* Compute a hash of a string */
+/* This simply computes a XOR of the first numbers */
+int nmbrHash(nmbrString *s)
+{
+  static long salt[] = { 4938, 48977, 6897, 7293, 2663, 7925, 2999, 12238,
+      40033, 14038, 10699, 29746, 56108, 34526, 63576, 52053, 61949, 41177, 43740, 22822
+ };
+  long i;
+  long hash = 0;
+  i = -1;
+  while (i < 13 && s[i] != -1) {

```

Comment:
I'll answer this one now as it is interesting, useful for the rest of the review... and definitely deserves more information in the comments!

Actually, indices -1, -2 and -3 _are_ valid and used in `nmbrString`:
- Index `-1` is the length of the number string. See `nmbrLen` (mmdata.c line 1023)
- Index `-2` is the allocated length, i.e. how many numbers are available totally (could be more than the actual current length of the string). See `nmbrAllocLen` (mmdata.c line 1032)
- Index `-3` is the location in `memUsedPool` (for memory management)

In this specific case, I wanted to include the length of the string in the hash, which I think makes sense but shall have been explained.

In any case, **you clearly spotted a mistake** here because in the salt, index -1 is clearly invalid!

Reply:
Any use of negative indices should be wrapped behind a macro to provide more safety and clarity. (That is, there should be a macro like `#define nmbrLen(p) p[-1]` where `p` has type `nmbrString` which is a typedef for `int*` or what have you.